### DISCUSSION TITLE
Chemprop FAQs

### DISCUSSION THREAD

**Discussion by JacksonBurns** [chemprop/chemprop#1042] [Announcements]
This discussion will collate some common questions and developer answers.

**Comment by JacksonBurns** [comment]
How are hyperparameters chosen for the model, and are there any specific methods for automatic tuning

**Reply by KnathanM** [reply]
The default model hyperparameters are on the smaller size to allow for quicker start up with Chemprop and often does well enough. Automatic tuning can be done using RayTune either in a [notebook](https://github.com/chemprop/chemprop/blob/main/examples/hpopting.ipynb) or from the command [line](https://chemprop.readthedocs.io/en/main/tutorial/cli/hpopt.html).

**Comment by JacksonBurns** [comment]
Are there guidelines for choosing optimal hyperparameters like optimization method, learning rate, batch size, or number of epochs?

**Reply by KnathanM** [reply]
This depends a lot on the dataset. We'd recommend our default search space for hyperparameter optimization which you can find [here](https://github.com/chemprop/chemprop/blob/dd8df8b7aecefb309e538b6e9664849f3224db8c/chemprop/cli/hpopt.py#L61-L75). 
Note that `Adam` is the optimizer used in Chemprop and this is [hard coded](https://github.com/chemprop/chemprop/blob/dd8df8b7aecefb309e538b6e9664849f3224db8c/chemprop/models/model.py#L219). 
Learning rate and batch size can both be optimized.
For number of epochs, I recommend that you keep training until the error on your validation set reaches a minimum.

**Comment by JacksonBurns** [comment]
Do you try to incorporate RDKit and deepchem descriptors with the MPNN model?

**Reply by JacksonBurns** [reply]
Chemprop v2.0.5+ supports appending the RDKit descriptors calculated via RDKit directly or via `descriptastorus` by specifying a `--molecule-featurizers rdkit_2d` or `v1_rdkit_2d_normalized`.

We do not support deepchem descriptors. Deepchem requires `sympy`, which only officially supports Python 3.8 to 3.10 - chemprop supports 3.11+

**Comment by JacksonBurns** [comment]
Are there recommended preprocessing steps for featurization (e.g., handling noisy data, standardization)?

**Reply by JacksonBurns** [reply]
Extensive data preprocessing is not included in Chemprop because it is highly problem-dependent. The only steps chemprop will take are to apply unit variance normal scaling to target values, as is standard for many machine learning applications. Noise, outlier removal, etc. are up to the user.

**Comment by JacksonBurns** [comment]
Can the model handle mixed datasets with categorical and continuous features?

**Reply by KnathanM** [reply]
The short answer is yes. You can pass both categorical and continuous features as extra atom. bond, or molecular features. For the CLI this is described [here](https://chemprop.readthedocs.io/en/main/tutorial/cli/train.html#additional-features). Your categorical features could either be one hot encoded or whole numbers. Default categorical features in Chemprop (like atomic number) are one hot encoded.

**Comment by JacksonBurns** [comment]
How customizable is the architecture of the MulticomponentMPNN or BondMessagePassing layers?

**Reply by KnathanM** [reply]
I recommend looking through [this part](https://chemprop.readthedocs.io/en/main/tutorial/python/index.html) of the docs under "Model Modules" for how to customize parts of the architecture.

**Comment by JacksonBurns** [comment]
Is there a way to add additional layers or modify the model architecture?

**Reply by KnathanM** [reply]
This would be a bit difficult, but not impossible. Chemprop models are `torch.nn.Module`s so you can use methods like `add_module()` ([reference](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.add_module)). But you would then need to override the `.fingerprint()` method of `MPNN` or `MulticomponentMPNN`, or if your modification are to the `predictor` you would need to override the `.forward()` method of your predictor (`RegressionFFN.forward()` for example). These methods determine the order that layers of the model are called.

**Comment by JacksonBurns** [comment]
How does the model handle multi-task learning? What additional configurations are required?

**Reply by JacksonBurns** [reply]
Multitask learning is accomplished by changing the output dimension of the `FFN` to match the number of tasks. See [this demo notebook](https://github.com/chemprop/chemprop/blob/main/examples/multi_task.ipynb).

**Comment by JacksonBurns** [comment]
How does chemprop handle missing data in the SMILES strings or target columns?

**Reply by JacksonBurns** [reply]
- Missing SMILES will raise an error and stop execution
 - Missing targets will be masked out during training, i.e. they will not count towards metrics or loss.

**Comment by JacksonBurns** [comment]
Can the user define custom evaluation metrics?

**Reply by KnathanM** [reply]
Yes if Chemprop is used in a notebook, but not from the command line. You can implement your own metric following the interface of the `Metric` class (see [here](https://github.com/chemprop/chemprop/blob/dd8df8b7aecefb309e538b6e9664849f3224db8c/chemprop/nn/metrics.py#L46)) and then pass an instance of that metric when creating the MPNN model. `metric_list = [MyMetric()]; models.MPNN(metrics=metric_list)`.
Note that in v2.1 we may switch over to using `torchmetrics`.

**Comment by JacksonBurns** [comment]
Is there built-in support for cross-validation or leave-one-out validation methods?

**Reply by JacksonBurns** [reply]
Chemprop v2.0 implements cross validation with a configurable number of folds (`-k/--num-folds`). Leave-one-out cross validation is not implemented directly, but can be achieved by specifying the number of folds to be equal to 1/# of points - this is _not_ reccomended, since this will take an enormous amount of time.

Chemprop v2.1+ implements replicates instead of cross validation (`--num-replicates`). This allows users to achieve the same values for estimated model performance as cross-validation with fewer actual executions.

**Comment by JacksonBurns** [comment]
Is data augmentation supported, and if so, what techniques are available?

**Reply by JacksonBurns** [reply]
No, chemprop does not implement any data augmentation methods. They must be implemented manually.

**Comment by JacksonBurns** [comment]
What are the best practices for the uses of the checkpoint file?

**Reply by JacksonBurns** [reply]
The checkpoint file can be used to restart training in case it is interrupted and to enable fine-tuning of a previously trained chemprop model. See [this demo notebook](https://github.com/chemprop/chemprop/blob/main/examples/predicting.ipynb) for an example of the former, and [this section of the documentation](https://chemprop.readthedocs.io/en/latest/tutorial/cli/train.html#pretraining-and-transfer-learning) for an example of doing the latter via the command line interface.

**Comment by JacksonBurns** [comment]
Can one feed an MPNN into a gradient-boosted tree method?

**Reply by KnathanM** [reply]
A Chemprop model (MPNN + FFN) is trained end to end so both the MPNN and FFN are updated together. The output of the MPNN should be a good representation of the molecule since the FFN can use it to get a property prediction. It isn't clear though if a tree method would do a better job at going from the output of the MPNN to a property prediction than the FFN that was trained with the MPNN. You probably shouldn't use an untrained MPNN to convert a molecule to a vector to then pass through a tree method, though I suppose you could try.

**Comment by TianYaoCSU** [comment]
Issue with TypeError in ChemProp Hyperparameter Optimization
I’m currently using ChemProp for hyperparameter optimization, and I encountered the following error during training:
TP = torch.cat(self.TP, dim=0).sum(0, keepdim=True)  
TypeError: cat() received an invalid combination of arguments - got (Tensor, dim=int), but expected one of:  
 * (tuple of Tensors tensors, int dim = 0, *, Tensor out = None)  
 * (tuple of Tensors tensors, name dim, *, Tensor out = None)  
This occurred while running the command:
css
chemprop hpopt -i F_reduced_labeled.csv -o hp_result -t classification --num-replicates 5 --metrics roc prc accuracy f1 binary-mcc --save-smiles-splits --show-individual-scores --pytorch-seed 42 --data-seed 42 --search-parameter-keywords depth ffn_num_layers dropout init_lr max_lr final_lr warmup_epochs batch_size --raytune-num-samples 50 --raytune-search-algorithm optuna --raytune-trial-scheduler AsyncHyperBand --raytune-num-cpus 8  
Could you please help me understand what might be causing this issue? Any guidance would be greatly appreciated!

Thank you for your assistance!

**Reply by JacksonBurns** [reply]
Please open an issue
